### PR TITLE
Add clamd s3 virus scanner

### PR DESCRIPTION
**Pull request by samuelhwilliams**
## Summary
A small script which streams objects from s3 to a clamd server.

## Ticket
https://trello.com/c/i30jDD69/317-automatically-virus-scan-uploaded-files-framework-agreements

### CODE REVIEW

File: scripts/virus-scan-s3-bucket.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+Requires a clamd backend to be provided (e.g. https://hub.docker.com/r/mkodockx/docker-clamav/)
+
+You'll need to configure the required AWS environment variables, normally either AWS_PROFILE, or
+AWS_ACCESS_KEY_ID & AWS_SECRET_ACCESS_KEY.
+
+Example:
+    ./scripts/virus-scan-s3-bucket --since 2018-01-01T00:00:00Z digitalmarketplace-documents-production-production
+"""
+
+import argparse
```
Comment (idavidmcdonald) [inline scripts/virus-scan-s3-bucket.py]: Not a fan of docopt like the rest of our scripts? I must say I often find it quite frustrating so interested to see how argparse plays out for you

File: scripts/virus-scan-s3-bucket.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
+#!/usr/bin/env python3
+"""Virus scan the contents of an S3 bucket, optionally from a given date onwards.
+Requires a clamd backend to be provided (e.g. https://hub.docker.com/r/mkodockx/docker-clamav/)
+
+You'll need to configure the required AWS environment variables, normally either AWS_PROFILE, or
+AWS_ACCESS_KEY_ID & AWS_SECRET_ACCESS_KEY.
+
+Example:
+    ./scripts/virus-scan-s3-bucket --since 2018-01-01T00:00:00Z digitalmarketplace-documents-production-production
```
Comment (idavidmcdonald) [inline scripts/virus-scan-s3-bucket.py]: It's somewhat bothering me that you have an optional argument before the required bucket argument. Could we swap these round so it looks pretty? (sorry)

File: scripts/virus-scan-s3-bucket.py (hunk 1 above)
Comment (idavidmcdonald) [reply scripts/virus-scan-s3-bucket.py]: That's docopt I find it frustrating by the way, not argparse (which I've never used)

**Comment by idavidmcdonald** [comment]
Oh and you've got a bunch of flake8 errors - enjoy :)

File: scripts/virus-scan-s3-bucket.py (hunk 2 above)
Comment (samuelhwilliams) [reply scripts/virus-scan-s3-bucket.py]: Interesting. I consider optional parameters after the required ones as ugly, for scripts. 🤔

File: scripts/virus-scan-s3-bucket.py (hunk 1 above)
Comment (samuelhwilliams) [reply scripts/virus-scan-s3-bucket.py]: I really don't like docopt for a number of reasons. I think argparse is cleaner to use, and it's faster at the command line IMO.

File: scripts/virus-scan-s3-bucket.py (hunk 2 above)
Comment (idavidmcdonald) [reply scripts/virus-scan-s3-bucket.py]: Ah yes, fair point. Let's keep it as is

**Comment by samuelhwilliams** [comment]
Flake8 fixed 👍

File: scripts/virus-scan-s3-bucket.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+                   help='Perform a dry run - will bypass downloading and scanning files from S3.')
+
+    args = a.parse_args()
+
+    if args.since and not args.since.tzinfo:
+        logger.error('You must supply a timezone-aware ISO8601 datetime string. You probably just need to append `Z`'
+                     'to the end of your datetime string. Example: 2018-01-01T12:00:00Z')
+        sys.exit(-1)
+
+    print(args)
```
Comment (risicle) [inline scripts/virus-scan-s3-bucket.py]: oop left this in

File: scripts/virus-scan-s3-bucket.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
+#!/usr/bin/env python3
+"""Virus scan the contents of an S3 bucket, optionally from a given date onwards.
+Requires a clamd backend to be provided (e.g. https://hub.docker.com/r/mkodockx/docker-clamav/)
```
Comment (risicle) [inline scripts/virus-scan-s3-bucket.py]: :see_no_evil: :hear_no_evil:

**Comment by Wynndow** [comment]
The readme in the scripts repo has a brief explanation of the scripts it contains. Might be worth adding this to it?

**Comment by samuelhwilliams** [comment]
@idavidmcdonald I will add some detail to the manual, yep. This might be a convenient way to do the virus scans in the future for frameworks until we get something completely integrated into our infrastructure. We should probably do some work to get this running on Jenkins though (later). I'd certainly recommend using this in the future as opposed to anything more manual.

@Wynndow added a bit to the readme.

**Pull request merged**